### PR TITLE
[WIP] GitHub authentication

### LINKED ISSUE #211: Require users to sign in with GitHub to create gists
GitHub are removing the ability to create anonymous gists: https://github.com/blog/2503-deprecation-notice-removing-anonymous-gist-creation

It's happening quite soon, so we'll need to implement a login flow as a matter of priority.

### PR DESCRIPTION
**Pull request by Rich-Harris**
This will fix #211. The codebase is in a bad state though, so I'm also tidying some stuff up to make that easier

### CODE REVIEW

File: routes/repl/_UserMenu.html [hunk 1]
```diff
@@ -0,0 +1,58 @@
+<div class='user' on:mouseenter='set({ showMenu: true })' on:mouseleave='set({ showMenu: false })'>
+	<span>{{$user.displayName}}</span>
```
Comment (Conduitry) [inline routes/repl/_UserMenu.html]: It's possible to have no displayName (I don't have one!) and so this should probably fall back to the username. Right now I'm seeing 'null' here.

**Comment by Conduitry** [comment]
Trying to load the SSR'd http://localhost:3000/guide directly is throwing an exception `ReferenceError: store is not defined`. The home, repl, and blog pages don't seem to suffer from this.

File: app/server.js [hunk 2]
```diff
@@ -1,15 +1,107 @@
...
-			routes
+			routes,
+			store: req => {
+				const user = req.session.passport && req.session.passport.user;
+
+				return new Store({
+					user: user && {
+						// strip access token
+						id: user.id,
+						name: user.name,
```
Comment (Conduitry) [inline app/server.js]: I think this should be `username: user.username`.

**Comment by Conduitry** [comment]
Saving/updating an existing gist looks like it _mostly_ works, but it does lose README.md, data.json, and meta.json.

File: routes/repl/_UserMenu.html (hunk 1 above)
Comment (Rich-Harris) [reply routes/repl/_UserMenu.html]: huh! fixed

File: app/server.js (hunk 2 above)
Comment (Rich-Harris) [reply app/server.js]: yep, you're right — fixed

**Comment by Rich-Harris** [comment]
Fixed all those points now, thanks @Conduitry.

Some fairly substantial changes in this PR:

* You need to be logged in with GitHub to save the REPL as a gist
* Saving overwrites a gist in place, if you own it. You can also fork a gist
* You can change the name of a gist
* Internally, the examples are no longer stored in the repo, but are *also* gists — we just have a mapping from example slug to gist ID. This simplifies things greatly
* The REPL has been broken up into smaller components — the codebase is slightly less hairy now
* Compilation is done in a worker, meaning your keys don't get sticky when you're editing a larger example (e.g. the Hacker News one)

**Pull request merged**